redash/models.py

```diff
@@ -96,6 +106,131 @@ def get(self, query_id):
+            # update the timestamp with the current time
+            user_detail.updated_at = utcnow()
+            # save to Redis
+            user_detail.save()
+        except ValueError:
+            user_detail = cls.create(
+                user_id=user_id,
+                updated_at=utcnow(),
+            )
+        return user_detail
+
+    @classmethod
+    def sync(cls, chunksize=1000):
+        """
+        Syncs user details to Postgres (to the JSON field User.details).
+        """
+        to_sync = {}
+        try:
+            for user_detail in cls.all():
+                to_sync[user_detail.user_id] = user_detail
+
+            user_ids = list(to_sync.keys())
+            if not user_ids:
+                return
+            logger.info(
+                'syncing users: %s',
+                ', '.join([str(uid) for uid in user_ids])
+            )
+            # get all SQLA users that need to be updated
+            users = User.query.filter(User.id.in_(user_ids))
+            for i, user in enumerate(users):
+                update = to_sync[user.id]
+                updated_at = getattr(update, 'updated_at', None)
+                if updated_at is None:
+                    continue
+                user.active_at = updated_at
+                # flush changes to the database after a certain
+                # number of items and extend the list of keys to
+                # stop sync in case of exceptions
+                if i % chunksize == 0:
+                    db.session.flush()
+            db.session.commit()
+        except DBAPIError:
+            # reset list of keys to stop sync
+            pass
+        finally:
+            user_ids = [str(user_id) for user_id in to_sync.keys()]
+            if user_ids:
+                logger.info(
+                    'Deleting temporary user details for users %s',
+                    ', '.join(user_ids)
+                )
+                delete_query = [
+                    UserDetail.user_id == str(user_id)
+                    for user_id in user_ids
+                ]
+                UserDetail.query_delete(reduce(or_, delete_query))
+
+
+class UserDetailsExtension(object):
+    """
+    A Flask extension to keep user details updates in Redis and
+    sync it periodically to the database (User.details).
+    """
+    def update_user_detail(self, sender, *args, **kwargs):
+        """
+        Used as a Flask request_started signal callback that adds
+        the current user's details to Redis
+        """
+        if (
+            current_user.get_id() and
+            current_user.is_authenticated and
+            not current_user.is_api_user()
+        ):
+            UserDetail.update(current_user.id)
+
+    def init_app(self, app):
+        """
+        Initializes this pseudo Flask extension with the given
+        Flask app.
+        """
+        request_started.connect(self.update_user_detail, app)
+
+
+user_details = UserDetailsExtension()
```

Comment:
Maybe be worth extracting all the user details related code into its own module? 

For a long time I wanted to change `redash.models` into a package and extract different modules from it. While this is to be of a task to fit into this pull request, maybe we can start by moving all the code from `redash/models.py` into the `redash/models/__init__.py` and then gradually move things into modules?

Reply:
I like that, lemme do that.

Reply:
I've done that now, and it feels good to do part of that. FWIW due to the import system I had to do a few more of the models that just the user details, to prevent circular imports. The majority of the models is still in redash/models/__init__.py though.